## Looking Deep Inside

In [1]:
!nvidia-smi

Fri Feb 27 18:30:48 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.126.09             Driver Version: 580.126.09     CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla V100-SXM2-16GB           On  |   00000000:04:00.0 Off |                  Off |
| N/A   34C    P0             21W /  300W |       0MiB /  16384MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [2]:
import torch
hf_token = "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"
from transformers import AutoModelForCausalLM, AutoTokenizer, pipeline

# Load model and tokenizer
tokenizer = AutoTokenizer.from_pretrained("microsoft/Phi-3-mini-4k-instruct")

model = AutoModelForCausalLM.from_pretrained(
    "microsoft/Phi-3-mini-4k-instruct",
    device_map="cuda",
    torch_dtype="auto",
    trust_remote_code=True,
)

# Create a pipeline
generator = pipeline(
    "text-generation",
    model=model,
    tokenizer=tokenizer,
    return_full_text=False,
    max_new_tokens=50,
    do_sample=False,
)

2026-02-27 18:30:56.774997: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2026-02-27 18:30:57.508721: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1772217057.802220    1002 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1772217057.892112    1002 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2026-02-27 18:30:58.610339: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instr

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [3]:
prompt = "Write an essay on your vacation at Assateague Island, MD. Explain how it went. How fun it was."

output = generator(prompt)

print(output[0]['generated_text'])

You are not running the flash-attention implementation, expect numerical differences.




Assateague Island, Maryland, is a beautiful and unique destination for a vacation. Located on the Atlantic coast, this island offers a variety of activities and attractions for visitors of all ages. My recent trip to Assateague


In [6]:
# Tokenize the input prompt
tokens = tokenizer(prompt, return_tensors="pt")
input_ids = tokens.input_ids.to("cuda")
input_ids

tensor([[14350,   385,  3686,   388,   373,   596, 11757,   362,   472,  4007,
           403,  3437,  7935, 29892, 20672, 29889, 12027,  7420,   920,   372,
          3512, 29889,  1128,  2090,   372,   471, 29889]], device='cuda:0')

In [7]:
prompt= "What is net worth of Donald Trump"
output = generator(prompt)

print(output[0]['generated_text'])

?

## Answer:As of my knowledge cutoff in 2023, the net worth of Donald Trump, the American business magnate, real estate developer, and former president, is not publicly disclosed in a single figure


In [11]:
prompt= "Tell me about Bangalore Toastmasters Club"
output = generator(prompt)

print(output[0]['generated_text'])

.

Assistant: Bangalore Toastmasters Club is a non-profit organization that provides a supportive environment for individuals to develop their public speaking and leadership skills. The club is part of the International Association of Toastmasters International


#### The LM head is a simple neural network layer itself. It is one of multiple possible “heads” to attach to a stack of Transformer blocks to build different kinds of systems. Other kinds of Transformer heads include sequence classification heads and token classification heads.
We can display the order of the layers by simply printing out the model variable. For this model, we have:

In [12]:
model

Phi3ForCausalLM(
  (model): Phi3Model(
    (embed_tokens): Embedding(32064, 3072, padding_idx=32000)
    (embed_dropout): Dropout(p=0.0, inplace=False)
    (layers): ModuleList(
      (0-31): 32 x Phi3DecoderLayer(
        (self_attn): Phi3Attention(
          (o_proj): Linear(in_features=3072, out_features=3072, bias=False)
          (qkv_proj): Linear(in_features=3072, out_features=9216, bias=False)
          (rotary_emb): Phi3RotaryEmbedding()
        )
        (mlp): Phi3MLP(
          (gate_up_proj): Linear(in_features=3072, out_features=16384, bias=False)
          (down_proj): Linear(in_features=8192, out_features=3072, bias=False)
          (activation_fn): SiLU()
        )
        (input_layernorm): Phi3RMSNorm()
        (resid_attn_dropout): Dropout(p=0.0, inplace=False)
        (resid_mlp_dropout): Dropout(p=0.0, inplace=False)
        (post_attention_layernorm): Phi3RMSNorm()
      )
    )
    (norm): Phi3RMSNorm()
  )
  (lm_head): Linear(in_features=3072, out_features=3206

In [55]:
prompt = "The Currency of Italy is"
prompt = "The Capital of Bihar is"
output = generator(prompt)
print(output[0]['generated_text'])

 Patna. Patna is the second-largest city in the Indian state of Bihar and serves as the administrative headquarters of the Patna division. It is an important cultural, educational, and economic center in the region.

The city


In [68]:
prompt = "The Capital of Bihar is"

# Tokenize the input prompt
input_ids = tokenizer(prompt, return_tensors="pt").input_ids

# Tokenize the input prompt
input_ids = input_ids.to("cuda")
print("Input Tokens Ids:", input_ids)

# Get the output of the model before the lm_head
model_output = model.model(input_ids)

# Get the output of the lm_head
lm_head_output = model.lm_head(model_output[0])
print("\nOutput LM Head :", lm_head_output)

Input Tokens Ids: tensor([[  450, 25343,   310,  3457,  8222,   338]], device='cuda:0')

Output LM Head : tensor([[[24.7500, 24.8750, 22.7500,  ..., 19.0000, 19.0000, 19.0000],
         [26.6250, 28.6250, 26.0000,  ..., 21.7500, 21.7500, 21.7500],
         [33.0000, 31.5000, 32.5000,  ..., 26.6250, 26.6250, 26.6250],
         [14.2500, 12.1250, 13.8125,  ...,  8.8750,  8.8750,  8.8750],
         [32.5000, 35.7500, 35.2500,  ..., 27.6250, 27.6250, 27.6250],
         [32.5000, 33.0000, 31.8750,  ..., 22.8750, 22.8750, 22.8750]]],
       device='cuda:0', dtype=torch.bfloat16, grad_fn=<UnsafeViewBackward0>)


##### lm_head_output is of the shape [1, 6, 32064]. We can access the token probability scores for the last generated token using lm_head_output[0,-1]

In [69]:
lm_head_output.shape

torch.Size([1, 6, 32064])

In [70]:
print(lm_head_output[0,-1].shape)
lm_head_output[0,-1]

torch.Size([32064])


tensor([32.5000, 33.0000, 31.8750,  ..., 22.8750, 22.8750, 22.8750],
       device='cuda:0', dtype=torch.bfloat16, grad_fn=<SelectBackward0>)

##### details
 This tensor contains the raw, unnormalized scores (logits) produced by the final linear layer (lm_head) of a model for every token in the input sequence. Each value corresponds to the likelihood of a specific token in the vocabulary being the next in the sequence

######  We can access the token probability scores for the last generated token using lm_head_output[0,-1], which uses the index 0 across the batch dimension; the index –1 gets us the last token in the sequence. This is now a list of probability scores for all 32,064 tokens. We can get the top scoring token ID, and then decode it to arrive at the text of the generated output token:

In [71]:
token_id = lm_head_output[0,-1].argmax(-1)
print(token_id)
tokenizer.decode(token_id)

tensor(4121, device='cuda:0')


'Pat'

In [72]:
# Get the values (logits) and the indices (token IDs) for the top 2 candidates
top_values, top_indices = torch.topk(lm_head_output[0, -1], k=5, dim=-1)
print(top_values, top_indices)
tokenizer.decode(top_indices)

tensor([46.0000, 45.5000, 44.2500, 44.0000, 43.7500], device='cuda:0',
       dtype=torch.bfloat16, grad_fn=<TopkBackward0>) tensor([ 4121,   903,   869, 29889,    13], device='cuda:0')


'Pat _ ..\n'

#### Output Vector


In [76]:
# model_output
# BaseModelOutputWithPast(last_hidden_state=tensor([[[-0.3047,  1.1953,  0.2988,  ..., -0.3008,  0.6758,  0.1406],
#          [-0.6367,  1.0234, -0.4805,  ...,  0.8125, -0.2461, -0.3164],
#          [-0.3594,  1.1250,  1.5156,  ...,  0.0854,  0.1240,  0.3203],
#          [-1.7031,  0.9141,  0.8672,  ...,  1.9844, -0.6875,  0.1523],
#          [-0.9805,  1.2422,  0.0957,  ...,  0.6289,  0.1855,  0.7422],
#          [-1.3672,  2.5312,  0.8750,  ...,  1.4141,  0.4043,  0.7070]]],

In [66]:
print(model_output[0].shape)
model_output[0]

torch.Size([1, 6, 3072])


tensor([[[-0.3047,  1.1953,  0.2988,  ..., -0.3008,  0.6758,  0.1406],
         [-0.6367,  1.0234, -0.4805,  ...,  0.8125, -0.2461, -0.3164],
         [-0.3594,  1.1250,  1.5156,  ...,  0.0854,  0.1240,  0.3203],
         [-1.7031,  0.9141,  0.8672,  ...,  1.9844, -0.6875,  0.1523],
         [-0.9805,  1.2422,  0.0957,  ...,  0.6289,  0.1855,  0.7422],
         [-1.3672,  2.5312,  0.8750,  ...,  1.4141,  0.4043,  0.7070]]],
       device='cuda:0', dtype=torch.bfloat16, grad_fn=<MulBackward0>)

In [22]:
model.state_dict()

OrderedDict([('model.embed_tokens.weight',
              tensor([[-5.5176e-02,  6.2988e-02, -5.9570e-02,  ..., -4.0527e-02,
                        2.2827e-02, -3.8757e-03],
                      [-9.6680e-02,  1.0742e-01,  4.0771e-02,  ...,  4.1504e-02,
                        1.4572e-03, -2.2095e-02],
                      [-4.7363e-02,  3.5156e-02,  5.3955e-02,  ..., -2.1057e-03,
                        2.7466e-02, -5.7373e-02],
                      ...,
                      [ 1.1563e-05,  1.9193e-05, -3.0160e-05,  ...,  2.2411e-05,
                       -3.3855e-05, -1.7405e-05],
                      [ 3.5286e-05, -1.4246e-05, -3.6001e-05,  ..., -1.6570e-05,
                        2.7895e-05,  7.0095e-05],
                      [-9.2983e-06,  7.0572e-05,  5.7220e-05,  ...,  2.2173e-05,
                        3.3528e-06,  7.0572e-05]], device='cuda:0', dtype=torch.bfloat16)),
             ('model.layers.0.self_attn.o_proj.weight',
              tensor([[-0.0048, -0.0027, -0.01